### PR TITLE
[shopsys] Webpack for less, svg

### PR DESCRIPTION
**Pull request by TomasGottvald**
| Q             | A
| ------------- | ---
|Description, reason for the PR| This PR replaces grunt with webpack for Frontend and administrations.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| Yes 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: project-base/webpack.config.js [hunk 1]
```diff
@@ -47,16 +47,146 @@ Encore
...
+                }
+                console.log('Frontend SVG icon ' + newFileFrontendPath + ' optimized');
+            });
+        });
+    });
+});
+
+const svgFilesAdmin = requireContext('../../../packages/framework/src/Resources/svg/admin', false, '.svg');
+var svgFilesAdminPath = [];
+svgFilesAdmin.keys().forEach((filePathAdmin) => {
```
Comment (pesektomas) [inline project-base/webpack.config.js]: I see violation DRY principle. I see the same block of code for frontend and admin. 

And I don't really like the location. I think this call should be linked to the webpack cycle. Check out EventHooksPlugin

File: project-base/webpack.config.js [hunk 2]
```diff
@@ -47,16 +47,146 @@ Encore
...
+Encore
+    .addEntry('admin-style', '../packages/framework/src/Resources/styles/admin/main.less')
+    .addPlugin(
+        new StylelintPlugin({
+            configFile: '.stylelintrc',
+            files: 'src/**/*.less'
+        })
+    )
+    .addLoader(
+        new WebfontsGenerator({
```
Comment (pesektomas) [inline project-base/webpack.config.js]: This is another DRY violation. This configuration is similar to WebfontsGenerator for admin. Maybe I recommend creating a new function that will return the configuration for WebfontsGenerator

File: project-base/webpack.config.js [hunk 3]
```diff
@@ -47,16 +47,146 @@ Encore
...
     }))
     .addPlugin(new CopyPlugin([
-        { from: 'web/bundles/fpjsformvalidator', to: '../../assets/js/bundles/fpjsformvalidator', force: true }
+        { from: 'web/bundles/fpjsformvalidator', to: '../../assets/js/bundles/fpjsformvalidator', force: true },
+        { from: 'web/bundles/shopsysframework', to: '../../web/assets/admin/images', force: true }
     ]))
 ;
 
+// Frontend Config
+const WebfontsGenerator = require('webfonts-generator');
```
Comment (pesektomas) [inline project-base/webpack.config.js]: You should place all require calls at the top of the file.

File: project-base/webpack.config.js [hunk 4]
```diff
@@ -47,16 +47,153 @@ Encore
...
+        { from: 'web/bundles/fpjsformvalidator', to: '../../assets/js/bundles/fpjsformvalidator', force: true },
+        { from: 'web/bundles/shopsysframework', to: '../../web/assets/admin/images', force: true }
     ]))
 ;
 
+
+// Frontend Config
+const WebfontsGenerator = require('webfonts-generator');
+const StylelintPlugin = require('stylelint-webpack-plugin');
+var requireContext = require('require-context');
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: Fixed

File: project-base/webpack.config.js [hunk 5]
```diff
@@ -47,16 +47,153 @@ Encore
...
+yaml = require('js-yaml');
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: fixed

File: project-base/webpack.config.js [hunk 6]
```diff
@@ -47,16 +47,153 @@ Encore
...
+fs = require('fs');
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: fixed

File: project-base/webpack.config.js [hunk 7]
```diff
@@ -47,16 +47,153 @@ Encore
...
+
+SVGO = require('svgo'),
+svgo = new SVGO();
+
+const svgFilesFrontend = requireContext('../../src/Resources/svg/front', false, '.svg');
+var svgFilesFrontendPath = [];
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: fixed

File: project-base/config/domains.yml [hunk 8]
```diff
@@ -2,6 +2,7 @@ domains:
     -   id: 1
         name: shopsys
         locale: en
+        styles_directory: common
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/config/domains.yml]: We need it for entry name - I thing it is better than if(styles_directory is null) condition.

File: project-base/package.json [hunk 9]
```diff
@@ -42,11 +42,18 @@
         "grunt-webfont": "^1.7.2",
         "jest": "^25.1.0",
         "jit-grunt": "^0.10.0",
+        "js-yaml": "^3.13.1",
+        "less-loader": "^4.1.0",
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/package.json]: Yes, It is my goal :D I feel this webpack configuration not completely done - due to admin paths, libs images. After resovling this I will remove all grunt things.

File: project-base/templates/Front/Layout/base.html.twig [hunk 10]
```diff
@@ -7,11 +7,12 @@
         <meta http-equiv="X-UA-Compatible" content="IE=11" />
         {% block meta_robots %}{% endblock %}
 
-        <link rel="stylesheet" type="text/css" href="{{ asset('assets/frontend/styles/index' ~ getDomain().id ~ '_' ~ getCssVersion() ~ '.css') }}" media="screen, projection">
-        <link rel="stylesheet" type="text/css" href="{{ asset('assets/frontend/styles/print' ~ getDomain().id ~ '_' ~ getCssVersion() ~ '.css') }}" media="print">
+        {% set entryDirectory = 'frontend-style-' ~ getDomain().currentDomainConfig.stylesDirectory %}
+        {{ encore_entry_link_tags( entryDirectory ) }}
 
-        {{ render(controller('App\\Controller\\Front\\HeurekaController:embedWidgetAction')) }}
+        <link rel="stylesheet" type="text/css" href="{{ asset('build/frontend-print-style-' ~ getDomain().currentDomainConfig.stylesDirectory ~ '.css') }}" media="print">
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/templates/Front/Layout/base.html.twig]: I tried it... but I did not find any way to add `media="print"` in it... do you know how to do that?

File: project-base/webpack.config.js [hunk 11]
```diff
@@ -47,16 +47,146 @@ Encore
...
 
+// Frontend Config
+const WebfontsGenerator = require('webfonts-generator');
+const StylelintPlugin = require('stylelint-webpack-plugin');
+var requireContext = require('require-context');
+yaml = require('js-yaml');
+fs = require('fs');
+
+SVGO = require('svgo');
+const svgo = new SVGO();
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: After changing this - it breaks svg.optimize ...

File: project-base/webpack.config.js [hunk 12]
```diff
@@ -47,16 +47,146 @@ Encore
...
+
+const svgFilesFrontend = requireContext('../../src/Resources/svg/front', false, '.svg');
+var svgFilesFrontendPath = [];
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: fixed

File: project-base/webpack.config.js [hunk 13]
```diff
@@ -47,16 +47,146 @@ Encore
...
+svgFilesFrontend.keys().forEach((filePathFrontend) => {
+    var newFileFrontendPath = './src/Resources/svg/front/' + filePathFrontend;
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/webpack.config.js]: fixed

File: project-base/webpack.config.js (hunk 1 above)
Comment (TomasGottvald) [reply project-base/webpack.config.js]: I know about this DRY - I feel this part as WIP - we need to check admin paths and libs images.

File: project-base/webpack.config.js (hunk 2 above)
Comment (TomasGottvald) [reply project-base/webpack.config.js]: I agree...

File: project-base/webpack.config.js (hunk 3 above)
Comment (TomasGottvald) [reply project-base/webpack.config.js]: Fixed

File: project-base/webpack.config.js (hunk 2 above)
Comment (TomasGottvald) [reply project-base/webpack.config.js]: This part was refactored and used as local function from assets/js/command file

File: project-base/webpack.config.js (hunk 1 above)
Comment (TomasGottvald) [reply project-base/webpack.config.js]: This part was refactored

File: project-base/src/Resources/styles/admin/main.less [hunk 14]
```diff
@@ -1,5 +1,5 @@
 // load main.less file from framework, variable frameworkResourcesDirectory is set in gruntfile.js
-@import "@{frameworkResourcesDirectory}/styles/admin/main.less";
+@import "../../../../packages/framework/src/Resources/styles/admin/main.less";
```
[the remark this answers is not in the document]
Comment (TomasGottvald) [reply project-base/src/Resources/styles/admin/main.less]: This part was refactored by you :)

File: project-base/.stylelintignore [hunk 15]
```diff
@@ -1,4 +1,4 @@
-src/**/helpers/*.*
-src/**/libs/*.*
-src/**/libs/**/*.*
-src/**/core/reset.less
+assets/styles/**/helpers/*.*
+assets/**/libs/*.*
+assets/**/libs/**/*.*
```
Comment (pesektomas) [inline project-base/.stylelintignore]: Is it necessary? I think so we don't have css by third path in project, or not?

File: project-base/assets/styles/admin/common/libs/svg.less [hunk 16]
```diff
@@ -0,0 +1,177 @@
+@font-face {
```
Comment (pesektomas) [inline project-base/assets/styles/admin/common/libs/svg.less]: should be this file in git?

File: project-base/web/public/frontend/styles/front/common/libs/svg.less [hunk 17]
```diff
@@ -0,0 +1,60 @@
+@font-face {
```
Comment (pesektomas) [inline project-base/web/public/frontend/styles/front/common/libs/svg.less]: should be this file in git? (maybe it is my mistake :smile:)

**Pull request merged**